winml/lib/Api/LearningModelSession.cpp

```diff
@@ -128,23 +129,29 @@
 
     com_ptr<winmlp::LearningModelSessionOptions> session_options_impl = session_options_.as<winmlp::LearningModelSessionOptions>();
 
-    // Make Onnxruntime apply the number of intra op threads
-    uint32_t numIntraOpThreads = session_options_impl->GetIntraOpNumThreads();
-    WINML_THROW_IF_FAILED(engine_builder->SetIntraOpNumThreadsOverride(numIntraOpThreads));
-    
     // Make onnxruntime apply named dimension overrides, if any
     if (session_options_impl && session_options_impl->NamedDimensionOverrides().Size() > 0) {
       WINML_THROW_IF_FAILED(engine_builder->SetNamedDimensionOverrides(session_options_impl->NamedDimensionOverrides()));
     }
-    bool allowSpinning = session_options_impl->GetIntraOpThreadSpinning();
-    WINML_THROW_IF_FAILED(engine_builder->SetIntraOpThreadSpinning(allowSpinning));
 
+    allow_spinning = session_options_impl->GetIntraOpThreadSpinning();
+    num_intra_op_threads = session_options_impl->GetIntraOpNumThreads();
+  }
+  
+  bool create_local_thread_pool = allow_spinning != device_impl->AllowSpinning() ||
+                                  num_intra_op_threads != device_impl->NumberOfIntraOpThreads();
+  if (create_local_thread_pool) {
+    WINML_THROW_IF_FAILED(engine_builder->SetIntraOpThreadSpinning(allow_spinning));
+    WINML_THROW_IF_FAILED(engine_builder->SetIntraOpNumThreadsOverride(num_intra_op_threads));
   } else {
-    // Onnxruntime will use half the number of concurrent threads supported on the system
-    // by default. This causes MLAS to not exercise every logical core.
-    // If session options aren't provided, force the thread pool size to be maxxed out
-    // to ensure that WinML always runs the fastest.
-    WINML_THROW_IF_FAILED(engine_builder->SetIntraOpNumThreadsOverride(std::thread::hardware_concurrency()));
+    winrt::com_ptr<_winml::IThreading> thread_pool = nullptr;
+    WINML_THROW_IF_FAILED(device_impl->GetThreadPool(thread_pool.put()));
+    if (thread_pool == nullptr)
+    {
+      WINML_THROW_IF_FAILED(engine_factory_->CreateThreadPool(allow_spinning, num_intra_op_threads, thread_pool.put()));
+      WINML_THROW_IF_FAILED(device_impl->CacheThreadPool(thread_pool.get()));
+    }
```

Comment:
The weirdness here is so that the shared thread pool on the device can be used to honor the session options (spinning, and intra op threads counts).

Ideally these would be optoins exposed on the device and not randomly picked up from the first session created with the device.

In this scheme the session options are always honored, but the shared thread pool on the device is configured based on the first session created.

The end user should never end up with behavior that they dont like, except possibly additional threads created when the internal device's thread pool was configured in an unexpected way (1st session's properties).

Since this behavior is undefined anyway, im okay with keeping this weird, as it provides some mechanism for exposing the spinning and num threads options to the device's shared thread pool (albeit unexpectedly).

A future API change is required to make the device's spinning and thread count properties available. This will allow the construction of the thread pool inside LearnignModelDevice and not as a side effect of Session creation.